### PR TITLE
Added buildradius checkbox to lobby options

### PR DESCRIPTION
**Pull request by ghost**
Implements #14203

### CODE REVIEW

**Comment by Kosti1** [comment]
sorry about dune2k commits, I do not know alot about it

File: OpenRA.Mods.Common/Traits/Buildings/Building.cs [hunk 1]
```diff
@@ -164,7 +164,7 @@ public Actor FindBaseProvider(World world, Player p, CPos topLeft)
 
 				// Range is counted from the center of the actor, not from each cell.
 				var target = Target.FromPos(bp.Actor.CenterPosition);
-				if (target.IsInRange(center, bp.Trait.Info.Range))
+				if (target.IsInRange(center, bp.Trait.Info.Range) || !world.WorldActor.Trait<MapBuildRadius>().BuildRadiusEnabled)
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/Buildings/Building.cs]: We already query the `MapBuildRadius` trait above and should 'cache' it therefore, imho.
I.e. something like:
```
public Actor FindBaseProvider(World world, Player p, CPos topLeft)
{
	var center = world.Map.CenterOfCell(topLeft) + CenterOffset(world);
	var mapBuildRadius = world.WorldActor.Trait<MapBuildRadius>();
	var allyBuildEnabled = mapBuildRadius.AllyBuildRadiusEnabled;

	foreach (var bp in world.ActorsWithTrait<BaseProvider>())
	{
		[...]

		if (mapBuildRadius.BuildRadiusEnabled || target.IsInRange(center, bp.Trait.Info.Range))
			return bp.Actor;
	}

	return null;
}
```

File: OpenRA.Mods.Common/Traits/World/MapBuildRadius.cs [hunk 2]
```diff
@@ -45,6 +54,8 @@ void INotifyCreated.Created(Actor self)
 		{
 			AllyBuildRadiusEnabled = self.World.LobbyInfo.GlobalSettings
 				.OptionOrDefault("allybuild", info.AllyBuildRadiusEnabled);
+			BuildRadiusEnabled = self.World.LobbyInfo.GlobalSettings
+        .OptionOrDefault("buildradius", info.BuildRadiusEnabled);
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/World/MapBuildRadius.cs]: Those spaces should be replaced by tabs.

File: OpenRA.Mods.Common/Traits/Buildings/BaseProvider.cs [hunk 3]
```diff
@@ -80,7 +80,7 @@ bool ValidRenderPlayer()
 		public IEnumerable<IRenderable> RangeCircleRenderables(WorldRenderer wr)
 		{
 			// Visible to player and allies
-			if (!ValidRenderPlayer())
+			if (!ValidRenderPlayer() || !wr.World.WorldActor.Trait<MapBuildRadius>().BuildRadiusEnabled)
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/Buildings/BaseProvider.cs]: I think we should cache `BuildRadiusEnabled` here as well as `AllyBuildEnabled` is already. (Also see the comment below.)

**Comment by abcdefg30** [comment]
> sorry about dune2k commits, I do not know alot about it

Don't worry. ^^ Just squash the commits together later on.
(Btw, if you got questions, feel free to reach out to me or someone else on our IRC channel [#openra](http://webchat.freenode.net/?channels=openra).)

**Comment by GraionDilach** [comment]
This is fundamentally wrong and suffers from tunnel vision - the ticket would be the perfect case to be resolved akin with a mutator, since you're now propagating a trait used in a single mod to be disabled in every shipped and thirdparty one (you can't really disable lobby checkboxes).

**Comment by Kosti1** [comment]
is tarvis-ci failing at getting mono usual?

**Comment by pchote** [comment]
> is tarvis-ci failing at getting mono usual?

No, this appears to be an upstream issue.  I assume that they should fix it within a day or two.
In any case, the tests would fail because you're still using space indentation in a few places.  Can you please fix that?

File: mods/common/chrome/lobby-options.yaml [hunk 4]
```diff
@@ -48,6 +48,11 @@ Background@LOBBY_OPTIONS_BIN:
 					Width: 150
 					Height: 20
 					Text: Debug Menu
+				Checkbox@BUILDRADIUS_CHECKBOX:
```
Comment (pchote) [inline mods/common/chrome/lobby-options.yaml]: Tiberian Sun doesn't include the base radius, so can you please duplicate the `lobby-options.yaml` into `mods/ts/chrome` and adjust its `mod.yaml` to use it, the same way d2k does?

File: mods/cnc/chrome/lobby-options.yaml [hunk 5]
```diff
@@ -53,6 +53,12 @@ Background@LOBBY_OPTIONS_BIN:
 					Height: 20
 					Font: Regular
 					Text: Debug Menu
+				Checkbox@BUILDRADIUS_CHECKBOX:
+					Y: 70
+					Width: 150
+					Height: 20
+					Font: Regular
+					Text: Build Radius
```
Comment (pchote) [inline mods/cnc/chrome/lobby-options.yaml]: "Build Radius" isn't a very obvious name.  Can we change this to something like "Limit ConYard Area"?

File: OpenRA.Mods.Common/Traits/Buildings/BaseProvider.cs [hunk 6]
```diff
@@ -46,6 +47,7 @@ public BaseProvider(Actor self, BaseProviderInfo info)
 			devMode = self.Owner.PlayerActor.Trait<DeveloperMode>();
 			progress = total = info.InitialDelay;
 			allyBuildEnabled = self.World.WorldActor.Trait<MapBuildRadius>().AllyBuildRadiusEnabled;
+            buildRadiusEnabled = self.World.WorldActor.Trait<MapBuildRadius>().BuildRadiusEnabled;
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Buildings/BaseProvider.cs]: Space indentation.

File: OpenRA.Mods.Common/Traits/Buildings/Building.cs [hunk 7]
```diff
@@ -154,7 +154,9 @@ public WVec CenterOffset(World w)
 		public Actor FindBaseProvider(World world, Player p, CPos topLeft)
 		{
 			var center = world.Map.CenterOfCell(topLeft) + CenterOffset(world);
-			var allyBuildEnabled = world.WorldActor.Trait<MapBuildRadius>().AllyBuildRadiusEnabled;
+            var mapBuildRadius = world.WorldActor.Trait<MapBuildRadius>();
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Buildings/Building.cs]: Space indentation.

File: OpenRA.Mods.Common/Traits/World/MapBuildRadius.cs [hunk 8]
```diff
@@ -45,6 +54,8 @@ void INotifyCreated.Created(Actor self)
 		{
 			AllyBuildRadiusEnabled = self.World.LobbyInfo.GlobalSettings
 				.OptionOrDefault("allybuild", info.AllyBuildRadiusEnabled);
+			BuildRadiusEnabled = self.World.LobbyInfo.GlobalSettings
+                .OptionOrDefault("buildradius", info.BuildRadiusEnabled);
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/World/MapBuildRadius.cs]: Space indentation.

**Review by pchote — DISMISSED** [review]

**Comment by pchote** [comment]
Looks good to me, otherwise.

**Comment by Kosti1** [comment]
fixing in few moments, there are no spacing problems on my local files

**Comment by pchote** [comment]
Sorry for the about face, but can you please duplicate this for ra and keep the common copy unmodified? As @graiondilach mentioned this probably doesnt want to be forced on downstream mods

**Comment by Kosti1** [comment]
@pchote done and working

File: OpenRA.Mods.Common/Traits/Buildings/Building.cs [hunk 9]
```diff
@@ -164,7 +166,7 @@ public Actor FindBaseProvider(World world, Player p, CPos topLeft)
 
 				// Range is counted from the center of the actor, not from each cell.
 				var target = Target.FromPos(bp.Actor.CenterPosition);
-				if (target.IsInRange(center, bp.Trait.Info.Range))
+				if (target.IsInRange(center, bp.Trait.Info.Range) || !buildRadiusEnabled)
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Buildings/Building.cs]: Please change this to return null before the loop if `!buildRadiusEnabled`, and then:

1. Query `MapBuildRadius` in the `Building` constructor and store the result in a `readonly bool buildRadiusEnabled`.
2. Check that value in `IsCloseEnoughToBase` to avoid calling this method entirely.

**Review by pchote — COMMENTED** [review]
Looks better, thanks.  There are a couple more minor changes you can do to improve performance:

File: OpenRA.Mods.Common/Traits/Buildings/BaseProvider.cs [hunk 10]
```diff
@@ -74,7 +78,7 @@ public bool Ready()
 
 		bool ValidRenderPlayer()
 		{
-			return self.Owner == self.World.RenderPlayer || (allyBuildEnabled && self.Owner.IsAlliedWith(self.World.RenderPlayer));
+			return buildRadiusEnabled ? self.Owner == self.World.RenderPlayer || (allyBuildEnabled && self.Owner.IsAlliedWith(self.World.RenderPlayer)) : false;
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/Buildings/BaseProvider.cs]: I think this can be simplified to `return buildRadiusEnabled && (self.Owner == self.World.RenderPlayer || (allyBuildEnabled && self.Owner.IsAlliedWith(self.World.RenderPlayer)));`.

File: OpenRA.Mods.Common/Traits/Buildings/BaseProvider.cs [hunk 11]
```diff
@@ -38,14 +39,17 @@ public class BaseProvider : ITick, INotifyCreated, IRenderAboveShroudWhenSelecte
 		int total;
 		int progress;
 		bool allyBuildEnabled;
+		bool buildRadiusEnabled;
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/Buildings/BaseProvider.cs]: Is there a reason why this is not `readonly`?

Comment (Kosti1) [reply OpenRA.Mods.Common/Traits/Buildings/BaseProvider.cs]: Its not because allyBuildEnabled is not

Comment (abcdefg30) [reply OpenRA.Mods.Common/Traits/Buildings/BaseProvider.cs]: Hm, I think you should make both `readonly` then. ^^
(They only get a value assigned in the constructor.)

**Comment by Kosti1** [comment]
Is this better?

File: OpenRA.Mods.Common/Traits/Buildings/BaseProvider.cs [hunk 12]
```diff
@@ -30,22 +30,26 @@ public class BaseProviderInfo : ITraitInfo
...
-		bool allyBuildEnabled;
 
 		public BaseProvider(Actor self, BaseProviderInfo info)
 		{
 			Info = info;
 			this.self = self;
 			devMode = self.Owner.PlayerActor.Trait<DeveloperMode>();
 			progress = total = info.InitialDelay;
-			allyBuildEnabled = self.World.WorldActor.Trait<MapBuildRadius>().AllyBuildRadiusEnabled;
+			mapBuildRadius = self.World.WorldActor.Trait<MapBuildRadius>();
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/Buildings/BaseProvider.cs]: Actually, we only use this in the constructor. So I'd suggest you make this a local `var` and drop the `readonly MapBuildRadius mapBuildRadius;`.

File: OpenRA.Mods.Common/Traits/Buildings/BaseProvider.cs [hunk 13]
```diff
@@ -74,8 +78,8 @@ public bool Ready()
 
 		bool ValidRenderPlayer()
 		{
-			return self.Owner == self.World.RenderPlayer || (allyBuildEnabled && self.Owner.IsAlliedWith(self.World.RenderPlayer));
-		}
+			return buildRadiusEnabled && (self.Owner == self.World.RenderPlayer || (allyBuildEnabled && self.Owner.IsAlliedWith(self.World.RenderPlayer)));
+        }
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/Buildings/BaseProvider.cs]: Somehow spaces sneaked in here again. Please change them to tabs.

**Review by abcdefg30 — COMMENTED** [review]
> Is this better?

Yes, just some more minor things and I hope we're done with changing the code back and forth. (Sorry for the hassle, btw...)

**Comment by Kosti1** [comment]
@abcdefg30 done, I hope there is nothing else (expecially spaces)

File: OpenRA.Mods.Common/Traits/Buildings/Building.cs [hunk 14]
```diff
@@ -154,7 +154,8 @@ public WVec CenterOffset(World w)
 		public Actor FindBaseProvider(World world, Player p, CPos topLeft)
 		{
 			var center = world.Map.CenterOfCell(topLeft) + CenterOffset(world);
-			var allyBuildEnabled = world.WorldActor.Trait<MapBuildRadius>().AllyBuildRadiusEnabled;
+			var mapBuildRadius = world.WorldActor.Trait<MapBuildRadius>();
+			var allyBuildEnabled = mapBuildRadius.AllyBuildRadiusEnabled;
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Buildings/Building.cs]: Please add an 
```csharp
if (!mapBuildRadius)
	return null;
```
as there are no base providers if the build radius is disabled.

**Review by pchote — COMMENTED** [review]
LGTM aside from one last minor nit:

File: OpenRA.Mods.Common/Traits/Buildings/Building.cs (hunk 14 above)
Comment (Kosti1) [reply OpenRA.Mods.Common/Traits/Buildings/Building.cs]: where exactly?

Comment (Kosti1) [reply OpenRA.Mods.Common/Traits/Buildings/Building.cs]: will do

**Comment by Kosti1** [comment]
@pchote is this good?

File: OpenRA.Mods.Common/Traits/Buildings/Building.cs [hunk 15]
```diff
@@ -154,7 +154,11 @@ public WVec CenterOffset(World w)
...
+
+            if (!mapBuildRadius.BuildRadiusEnabled)
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Buildings/Building.cs]: These lines are now indented using spaces.  Please replace them with tabs.

https://github.com/OpenRA/OpenRA/wiki/Coding-Standard includes a guide on how to configure Visual Studio to use tabs.

Comment (Kosti1) [reply OpenRA.Mods.Common/Traits/Buildings/Building.cs]: but I use linux

**Comment by Kosti1** [comment]
well great

**Review by pchote — APPROVED** [review]

**Review by NethIafin — APPROVED** [review]
Not a fan of this new row while there is just one item, but this works as advertised. LGTM

**Comment by pchote** [comment]
> Not a fan of this new row while there is just one item

There are two other long thought about checkboxes that could be implemented by somebody to fill in the gaps:
* Kill bounties
* Undeploy MCV.

**Comment by AndriiYukhymchak** [comment]
Yeah, got it. Maybe @kosti1 can create separate pr for those, so in playtest we would have full row. Either way this is good as it is

**Comment by Kosti1** [comment]
will look in to it

**Review by MustaphaTR — APPROVED** [review]
Looks good to me 👍

**Review by abcdefg30 — APPROVED** [review]

**Comment by abcdefg30** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/6240111f18e97457dacc565eeac5bdc97b83150c...9502b2c7987093fd72f32c4c1a07f109f413ed03)

**Comment by GraionDilach** [comment]
Congratulations to everyone for creating the worst precedent in regarding to customized lobby checkboxes! @pchote, since you were interested on the other day why I believe taht modder feedback is ignored in eaxact cases - here is one.

The CnCNetv5 client used by many TS/RA2 mods apply a rules edit behind the checkbox between the main rules and the map edits. That system should have been used here, not expanding the system with made-up arbitrary stuff.

**Comment by pchote** [comment]
@GraionDilach: I don't really understand your hostility towards this.  This PR doesn't set a precident; it follows our conventions that were established in #11364 and prerequisite PRs.  Asking a new contributor to implement a new low-level engine feature really doesn't make sense.

**Comment by GraionDilach** [comment]
What does this PR does in effect on RA rules? Let me sum it up with two lines.

```
FACT:
	-BaseProvider:
```

Where this PR does set a precedent is that lousy hardcoded lobby checkboxes which should have been implemented via custom rules are acceptable as C# code fragments and encouraged to do so. This precedent is already utilized at #14245.

There wasn't a need to completely change the lobby options besides providing a secondary method. #9422 was filed _exactly_ for lobby options causing this level of YAML edits. Ignoring it is convenient, blowing it to out-of-proportion levels is even more convenient - especially when one decides to ignore how CnCNet TS/YR is able to utilize said system for years now.

**Comment by penev92** [comment]
### OK, as for the help/feedback/attitude:
> since you were interested on the other day why I believe taht modder feedback is ignored in eaxact cases - here is one.

I checked - all you said prior to merge is "this is bad". That's all the feedback.
Now, assuming you had linked to the ticket that you opened (which you didn't, so ... shame on everyone else for not making the connection?), that ticket still goes exactly ankle-deep in the issue - it more or less just says "hey guys, CnCNet use X, UT2004 uses Y; they look like this: <*screenshots*>". You scratch the surface of an actual explanation of implementation details but go no further.
In fact, _the first time_ I can find you mentioning that you have an idea about _how_ to implement it other than visually was yesterday on Discord **after** this PR was merged. Now **that** was some decent feedback, but perhaps just a tiny a bit late, no?

### Now, as for the suggested implementation:
From what I've read that method sounds utterly unsupportable for more than 1 or 2 **very simple** mutators.
I don't know how CnCNet does it, maybe the inis there are better suited for the task, maybe it's a complete mess... but for the simplest of mutators you can easily need to have a copy of your mod's entire `rules` folder, just because it affects every actor you have.
So if you have 20 (again, **simple** mutators), you either have this folder setup in your mod:
```
rules_default
rules_mutator_1
rules_mutator_2
rules_mutator_3
...
rules_mutator_20
```
or you have these yamls in your rules folder along with all the others:
```
mutator1.yaml
mutator2.yaml
mutator3.yaml
...
mutator20.yaml
```
with each of those potentially having a 2-line entry (at best) for **every actor in the mod**.
I think you can see how that is going to become utterly unsupportable, unmaintainable, unmanageable...
Yes, it would be (somewhat) flexible (as long as you only want to tweak trait values, which is somewhat limited for a mutator), but the YAML overhead would be enough to completely negate any usefulness IMO.

### In conclusion,
1. I'm not saying the used approach is very good or scalable, but IMO neither is the one you're suggesting.
and
2. Please don't go all "I tried to help but they ignored my feedback so * them" again, since like I said, your feedback actually came _**after**_ this was merged.

So how about we all do something constructive for a change and try to come up with an actually good approach together?

**Comment by MustaphaTR** [comment]
> I think you can see how that is going to become utterly unsupportable, unmaintainable, unmanageable...

Sorry, but i can't. I agree with @GraionDilach that Mutators as CnCNet has would be way better alternative to this. I see no problem with having 20 mutator files, probably under rules/mutators folder not directly rules.

**Comment by GraionDilach** [comment]
I left a comment somewhere at @Kosti1's repo on a commit weeks ago explaining this exact same thing I'm atm pissed on but apparently it was lost in a rebase.
